### PR TITLE
optimize bucket level monitor to resolve alias to query only those time-series indices that contain docs within timeframe of range query filter in search input

### PR DESCRIPTION
**Pull request by eirsep**
### Description
optimize bucket level monitor to resolve alias to query only those time-series indices that contain docs within timeframe of range query filter in search input
### Related Issues
#1710 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/alerting/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: alerting/src/main/kotlin/org/opensearch/alerting/InputService.kt [hunk 1]
```diff
@@ -289,4 +358,53 @@ class InputService(
         }
         return results
     }
+
+    fun resolveStartTimeofQueryTimeRange(monitor: Monitor, query: QueryBuilder, periodEnd: Instant): Instant? {
+        try {
+            val rangeQuery = findRangeQuery(query) ?: return null
```
[the remark this answers is not in the document]
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/InputService.kt:364]: from both security analytics timeframe perspective and our own user experience we allow only one time range query which builds on `period_end` search param

[the remark this answers is not in the document]
Comment (AWSHurneyt) [reply alerting/src/main/kotlin/org/opensearch/alerting/InputService.kt:364]: If a user defines a monitor using the extraction query editor UX though, couldn't they have more than 1 range query?

[the remark this answers is not in the document]
Comment (eirsep) [reply alerting/src/main/kotlin/org/opensearch/alerting/InputService.kt:364]: here we are checking for time range queries only for performing this optimization

i am not sure how there can be multiple time ranges being queried upon. Any range query without the search param `period_end` is anyway not being considered

**Review by AWSHurneyt — APPROVED** [review]
Changes LGTM. Approving for when the separate PRs to fix the build error is merged.

File: alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt [hunk 2]
```diff
@@ -1190,6 +1191,133 @@ class MonitorRunnerServiceIT : AlertingRestTestCase() {
         assertEquals("Incorrect search result", 2, buckets.size)
     }
 
+    fun `test execute bucket-level monitor with alias optimization - indices not skipped`() {
```
Comment (sbcd90) [inline alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt:1194]: can we add a test case with partial index consideration e.g. after the current monitor run, `5 new docs are added to index1`, `index1 is rolled over to index2`, `5 new docs added to index2`, `index2 is rolled over to index3`, `5 new  docs added to index3`,  `index3 is rolled over to index4`, `5 new  docs added to index4` & now `bucket-level monitor run` happens. So, this run considers, `index1, index2, index3 & index4`.

Comment (eirsep) [reply alerting/src/test/kotlin/org/opensearch/alerting/MonitorRunnerServiceIT.kt]: added 2 test scenarios

scenario one 
12 indices in alias
1st index skipped as creation time out of time range of query contains 3 docs (3 new buckets in response)
2nd index NOT SKIPPED with creation time out of time range of query contains 3 docs (3 new buckets in response)
10 indices contains 2 docs (2 new buckets in response)
Response has only 5 buckets 

scenario 2
12 indices in alias
1st index NOT SKIPPED as creation time in time range of query contains 3 docs (3 new buckets in response)
2nd index NOT SKIPPED with creation time in time range of query contains 3 docs (3 new buckets in response)
10 indices contains 2 docs (2 new buckets in response)
Response has only 8 buckets

**Review by sbcd90 — APPROVED** [review]

**Review by AWSHurneyt — APPROVED** [review]

**Pull request merged**